library/sha512.c

```diff
@@ -188,9 +304,249 @@ static const uint64_t K[80] =
     UL64(0x4CC5D4BECB3E42B6),  UL64(0x597F299CFC657E2A),
     UL64(0x5FCB6FAB3AD6FAEC),  UL64(0x6C44198C4A475817)
 };
+#endif
 
-int mbedtls_internal_sha512_process( mbedtls_sha512_context *ctx,
-                                     const unsigned char data[128] )
+#if defined(MBEDTLS_SHA512_USE_A64_CRYPTO_IF_PRESENT) || \
+    defined(MBEDTLS_SHA512_USE_A64_CRYPTO_ONLY)
+
+#if defined(MBEDTLS_SHA512_USE_A64_CRYPTO_ONLY)
+#  define mbedtls_internal_sha512_process_many_a64_crypto mbedtls_internal_sha512_process_many
+#  define mbedtls_internal_sha512_process_a64_crypto      mbedtls_internal_sha512_process
+#endif
+
+#ifndef asm
+#define asm __asm__
+#endif
+
+/* Accelerated SHA-512 implementation originally written by Simon Tatham for PuTTY,
+ * under the MIT licence; dual-licensed as Apache 2 with his kind permission.
+ */
+
+#if defined(__clang__) && \
+           (__clang_major__ < 13 || \
+           (__clang_major__ == 13 && __clang_minor__ == 0 && __clang_patchlevel__ == 0))
+static inline uint64x2_t vsha512su0q_u64(uint64x2_t x, uint64x2_t y)
```

Comment:
Is this condition the only one that doesn't support intrisic?

Reply:
When I was doing this in PuTTY, clang pre-13 was the only compiler I knew of which didn't support the SHA-512 intrinsics but _did_ support the underlying machine instructions. (Obviously, plenty of compilers support neither, in which case even this workaround wouldn't work!)

In current PuTTY I've replaced this kind of check with a cmake test at configure time, which ought to detect any compiler in which these workaround implementations are both usable and needed. But that was just "on general principles" – I don't know of any specific compiler that would have been left out by the simple ifdefs.

Reply:
@xkqian In my testing, `clang` < 13 supported the instruction but not the intrinsics (as Simon says), and clang 13 supported the intrinsics. however, we found that clang == 13.0.0 behaved as clang < 13 (i.e. the intrinsics support was added in 13.0.1, which is very strange, but what we found)

Reply:
(and note that Mbed TLS supports `Makefile` builds as well as a `cmake` build, hence not building detection into the build system)